lib/join/circleci/circleci.go

```diff
@@ -0,0 +1,162 @@
+/*
+ * Teleport
+ * Copyright (C) 2023  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// CircleCI workload identity
+//
+// Key values:
+// Issuer of tokens: `https://oidc.circleci.com/org/ORGANIZATION_ID`
+// Audience: `ORGANIZATION_ID`
+//
+// `iat` and `exp` will be included and should be respected.
+//
+// Useful references:
+// - https://circleci.com/docs/openid-connect-tokens/
+
+package circleci
+
+import (
+	"context"
+	"fmt"
+	"slices"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/zitadel/oidc/v3/pkg/oidc"
+
+	"github.com/gravitational/teleport"
+	workloadidentityv1pb "github.com/gravitational/teleport/api/gen/proto/go/teleport/workloadidentity/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/join/provision"
+	logutils "github.com/gravitational/teleport/lib/utils/log"
+)
+
+var log = logutils.NewPackageLogger(teleport.ComponentKey, "circleci")
+
+// Validator is a function that can be used to validate CircleCI tokens
+type Validator func(ctx context.Context, organizationID, token string) (*IDTokenClaims, error)
+
+const IssuerURLTemplate = "https://oidc.circleci.com/org/%s"
+
+func issuerURL(template, organizationID string) string {
+	return fmt.Sprintf(template, organizationID)
```

Comment:
kind of a nit but this is just string concatenation, no need to Sprintf here

However, we might want to be a bit more careful to avoid potential SSRF, even though organizationID comes from the token. Could build the URL using the net/url package instead, maybe something like

```go
url.URL{
	Scheme: "https",
	Host: "oidc.circleci.com",
}.JoinPath("org", organizationID).String()
```